Send disc CRC with compatibility reports

unknownbrackets opened the pull request:
From published sequential read statistics, it looks like the vast majority of Android devices will be able to do a full read of any UMD-size ISO in <= 30 seconds (and the slowest devices, < 90s.)  This performs that operation on a background thread.

For the disk and memory caches (i.e. HTTP), it'll be slower, but I added a flag to avoid letting it throw away good cache data.

I do have a bit of a catch-22 here though.  Ideally I'd like to notify the user if their crc is bad.  Basically, I see people reporting games I have and know work (and also games marked as Perfect by others on the same devices) as "Doesn't boot", which either means they're submitting garbage, have some horribly bad settings, or have a corrupted ISO.

Perhaps if they submit poor feedback (e.g. < ingame, or <= ingame?), we could make them wait for the submission to complete, and then show some result?  Alternatively, it could start CRCing as soon as they open the report screen, and interrupt them on that screen with the results.  Not sure.  Neither option sounds great.

Any result could also contain recommended settings changes based on other, better reports.

Note that this also allows differentiating things like fan translations, which typically don't change the game id.

-[Unknown]

unknownbrackets committed:
Add uncached flag for file loader reads.

This allows us to indicate that a read need not be cached.

unknownbrackets committed:
Send a CRC hash with compatibility reports.

hrydgard commented:
Hm, tricky one. Reading flash consumes a lot of battery, so even if there's a spare CPU to use, it's not free, and I don't think this should be enabled by default, at least if not currently being charged from a power cable. But yes, there are definitely a ton of corrupted or ripped-for-size ISOs floating around out there, and it would be nice to be able to warn the user for ones that are often reported as not working - though of course, we would prefer users to dump games properly themselves.

unknownbrackets commented:
Hmm, we could add a checkbox (like the screenshot one) to calculate a CRC on `MOBILE_DEVICE`?  Note that right now - this is not for log reports - only for explicit user-initiated compatibility reports.

I suppose we could probably even add a System_GetProperty type thing for charging from cable, I guess?

Well, moreover we could tell the user "your disc's CRC doesn't match what users commonly report as working, this may mean your disc is corrupt."  SD cards can go bad, even when users have ripped themselves.  I'm not specifically after targetting "known bad" CRCs, as much as "not known good".  Of course, if a translation is reported as "good" then it can be on the whitelist.

-[Unknown]

unknownbrackets committed:
Show feedback when the compatibility is submitted.

unknownbrackets committed:
Reset flag when trying server anew.

unknownbrackets commented:
Okay, added a mobile-only checkbox with warning about battery impact.

Also, added a message when data submission (including CRC etc.) is complete.  This will also let you know if your report was rejected due to an ancient/blacklisted PPSSPP version, or if you otherwise have network issues.

-[Unknown]

unknownbrackets committed:
Stop reporting FrameRate.

This is just a speed setting, it doesn't affect emulation.

unknownbrackets commented:
What do you think about the checkbox?

-[Unknown]

hrydgard commented:
Hm, I think I misunderstood this feature initially. Should have actually read the code properly. I thought it would always compute a CRC when booting a game, and would simply submit it directly when reporting, instead of computing it and then submitting, when the user actually chooses to submit.

As this is only done when reporting, a simple warning for mobile users is probably enough, no checkbox required. Sorry for the trouble..

(also, needs a rebase)

unknownbrackets committed:
Add notice about battery-consuming CRC.

unknownbrackets commented:
Rebased 1cdb7a69ccd10fef8db69056e7a3287c77aa0340 out and added a small notice.

-[Unknown]

unknownbrackets commented:
Just a note: 34 people have reported compatibility with this merged feature.

Of those, 7 (21%) reported the game didn't go in game.  6 of those 7 (86%) had a bad CRC.  The other one was [on iOS](http://report.ppsspp.org/game/NPJH50838_1.00) and was using non-buffered rendering + frameskip 5 + disable slow effects + non-IR interpreter + etc.

-[Unknown]

unknownbrackets commented:
Added some crappy UI on the site to show this.  For example here:
http://report.ppsspp.org/game/NPJH50686_1.00

People often report problems with CRC a508d796, but not with CRC 32becc50.

No feedback within PPSSPP yet, though.

-[Unknown]
